src/main/scala/firrtl/Utils.scala

```diff
@@ -272,199 +243,138 @@ object Utils extends LazyLogging {
-         case _ => UnknownType
-       }
+     def getType(): Type = node match {
+       case e: Expression => e.getType
+       case s: Statement => s.getType
+       //case f: Field => f.getType
+       case t: Type => t.getType
+       case p: Port => p.getType
+       case _ => UnknownType
+     }
    }
 
 //// =============== EXPANSION FUNCTIONS ================
-   def get_size (t:Type) : Int = {
-      t match {
-         case (t:BundleType) => {
-            var sum = 0
-            for (f <- t.fields) {
-               sum = sum + get_size(f.tpe)
-            }
-            sum
-         }
-         case (t:VectorType) => t.size * get_size(t.tpe)
-         case (t) => 1
-      }
+   def get_size (t:Type) : Int = t match {
+      case (t:BundleType) => (t.fields foldLeft 0)(
+        (sum, f) => sum + get_size(f.tpe))
+      case (t:VectorType) => t.size * get_size(t.tpe)
+      case (t) => 1
    }
    def get_valid_points (t1:Type, t2:Type, flip1:Orientation, flip2:Orientation) : Seq[(Int,Int)] = {
       //;println_all(["Inside with t1:" t1 ",t2:" t2 ",f1:" flip1 ",f2:" flip2])
       (t1,t2) match {
          case (t1:UIntType,t2:UIntType) => if (flip1 == flip2) Seq((0, 0)) else Seq()
          case (t1:SIntType,t2:SIntType) => if (flip1 == flip2) Seq((0, 0)) else Seq()
-         case (t1:BundleType,t2:BundleType) => {
-            val points = ArrayBuffer[(Int,Int)]()
-            var ilen = 0
-            var jlen = 0
-            for (i <- 0 until t1.fields.size) {
-               for (j <- 0 until t2.fields.size) {
-                  val f1 = t1.fields(i)
-                  val f2 = t2.fields(j)
-                  if (f1.name == f2.name) {
-                     val ls = get_valid_points(f1.tpe,f2.tpe,times(flip1, f1.flip),times(flip2, f2.flip))
-                     for (x <- ls) {
-                        points += ((x._1 + ilen, x._2 + jlen))
-                     }
-                  }
-                  jlen = jlen + get_size(t2.fields(j).tpe)
-               }
-               ilen = ilen + get_size(t1.fields(i).tpe)
-               jlen = 0
-            }
-            points
-         }
-         case (t1:VectorType,t2:VectorType) => {
-            val points = ArrayBuffer[(Int,Int)]()
-            var ilen = 0
-            var jlen = 0
-            for (i <- 0 until scala.math.min(t1.size,t2.size)) {
+         case (t1:BundleType,t2:BundleType) => (
+           (t1.fields foldLeft (Seq[(Int, Int)](), 0)){case ((points, ilen), f1) => (points ++ (
+             (t2.fields foldLeft (Seq[(Int, Int)](), 0)){case ((points, jlen), f2) => (points ++ (
+               if (f1.name == f2.name) {
+                 val ls: Seq[(Int, Int)] = get_valid_points(f1.tpe,f2.tpe,times(flip1,f1.flip),times(flip2,f2.flip))
+                 ls map {case (x, y) => (x + ilen, y + jlen)} 
+               } else Nil
+             ), jlen + get_size(f2.tpe))}
+           )._1, ilen + get_size(f1.tpe))}
+         )._1
```

Comment:
This is really hard to understand. Does it actually provide a performance improvement over what was there before because I'm pretty lost